rfcs/0010-references.md

```diff
@@ -37,62 +37,119 @@ In most cases, you shouldn't need to use explicit pointer types.
+
+r_s.x = 1; // Same as `(*r_s).x = 1`.
 ```
 
-and produces the desired memory representation.
+Here, we are listing the major properties of references:
+- References have their own mutability and can point to mutable or immutable values.
+- & operator defines the reference. * operator is dereferencing.
+- . and [] operators also dereference if the reference is a reference to a struct/tuple or array, respectively.
+- References can be parts of aggregates.
+- References can reference other references.
+- References can be used in pattern matching and deconstructing.
+- References can reference parts of aggregates. E.g., having a reference to an array element.
+- References can be passed to and returned from functions.
+- References will play well with iterators and the `for` loop, once implemented.
+- References can be used together with generics.
+- References can be taken from arbitrary expressions, including constants and literals.
+- References can be used in type aliases.
+- `self` keyword will become a reference, and us such comply to the reference passing syntax.
+- References cannot be used in storage.
+- References cannot be used in ABIs and `main` functions.
+- Equality of references is the equality of the values they refer to if the underlying type implements `std::ops::Eq`.
+- `__addr_of` called on a reference returns the address the reference points to.
+
+For detailed examples of syntax and semantics see the accompanied file [0010-references.sw](../files/0010-references.sw).
 
 ## Pointers
 
 There are cases where one may want to directly manipulate memory addresses with pointer arithmetic,
-to allow for this we have a pointer type that represents a single address: `*mut T` where T is the type being pointed to.
+to allow for this we have a pointer type that represents a single address: `*mut T` where `T` is the type being pointed to.
 
 Pointers can be obtained by using the `__addr_of` intrinsic on a reference and dereferenced using the `__deref` intrinsic, like so:
 
 ```sway
 let val: u64 = 1;
-let ptr: *mut u64 = __addr_of(Box::new(val));
+let ptr: *mut u64 = __addr_of(&val);
 let ptr_val: u64 = __deref(ptr);
 assert_eq(val, ptr_val);
 
-let val: Box<u64> = Box::new(1);
-let ptr: *mut Box<u64> = __addr_of(Box::new(val));
-let ptr_val: Box<u64> = __deref(ptr);
-assert_eq(val, ptr_val);
+let ref: &u64 = &1;
+let ptr: *mut &u64 = __addr_of(&ref);
+let ptr_val: &u64 = __deref(ptr);
+assert_eq(ref, ptr_val); // Equality of references is the equality of referenced values.
 
-let val: Box<u64> = Box::new(1);
-let ptr: *mut u64 = __addr_of(val);
+let ref: &u64 = &1;
+let ptr: *mut u64 = __addr_of(ref); // Returns the address od the referenced value.
 let ptr_val: u64 = __deref(ptr);
-assert_eq(val, Box::new(ptr_val));
+assert_eq(ref, &ptr_val); // Equality of references is the equality of referenced values.
 ```
 
 Dereferencing an invalid pointer is Undefined Behavior.
```

Comment:
But can we guarantee that every deference of invalid pointers will panic? It is interesting to notice that there is no definition of invalid pointers. 

By `Every time the VM runs, a single monolithic memory of size VM_MAX_RAM bytes is allocated` and `The stack begins at 0 and grows upward. The heap begins at VM_MAX_RAM and grows downward.
` (https://specs.fuel.network/master/fuel-vm/index.html)
and the fact that loads only panics when they go beyond VM_MAX_RAM; we have that deferencing invalid pointers  will only revert with overflowing indices and pointer arithmatic. All others cases will read garbage and not revert. 

Is that right?

Reply:
@jjcnn @xunilrj Good points and valid concerns. I've added a definition of an invalid pointer and gave an explanation of behavior for FuelVM.

In the case of FuelVM the [access rights](https://specs.fuel.network/master/fuel-vm/index.html#access-rights) will ensure instructions revert in case of a memory access outside of the allocated area. But this is just FuelVM behavior. Once we start targeting other VMs this behavior might be different.

To the general question of runtime checks imposed by the Sway compiler, if a discussion is needed, let's make it as a separate amendment. The above statement on dereferencing invalid pointers was in the accepted part of the original RFC. It wasn't changed by this amendment that focus only on references (`&T`). If there are alternative ideas and concerns, let's rather have and additional amendment that will focus only on pointers. At the offsite last year we were discussing reference tracking and counting and are aware of trade offs between undefined and runtime-checked behavior. Also, note that we have similar tradeoffs in other areas. E.g., we opted for encoding format that does not carry any meta information.

Reply:
> I guess it depends whether we can determine the length of arrays.

In case of `[T;n]` arrays we already have compile-time checks for out of bounds access. But of course they work only for const-eval indices.